### PR TITLE
TravisCI + PHPUnit

### PR DESCRIPTION
**Pull request by iam-merlin**
#26 

With an intensive test suite... ><

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+script:
+  # this run phpunit/tests
+  - ./vendor/bin/phpunit
+  # this checks that the source code follows Code Syntax rules
+  # @todo update for Symfony v4
+  #- ./vendor/bin/phpcs --bootstrap=.php_cs
+
+  # this checks that the composer.json and composer.lock files are valid
+  - composer validate --strict
```
Comment (xelan) [inline .travis.yml]: Please add a newline at the end of configuration and PHP files to make viewing/editing files in the CLI easier.

File: tests/Helpers/UtilTest.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+namespace Resque\Tests\Heplers\Utils;
+
+use Resque\Helpers\Util;
+use PHPUnit\Framework\TestCase;
+
+class SluggerTest extends TestCase
+{
+    /**
+     * @param $expected
```
Comment (xelan) [inline tests/Helpers/UtilTest.php]: IMHO all of the parameters should have their types documented.

File: tests/Helpers/UtilTest.php [hunk 3]
```diff
@@ -0,0 +1,51 @@
+<?php
+
+namespace Resque\Tests\Heplers\Utils;
+
+use Resque\Helpers\Util;
+use PHPUnit\Framework\TestCase;
+
+class SluggerTest extends TestCase
```
Comment (xelan) [inline tests/Helpers/UtilTest.php]: Namespace and class name do not match the actual structure.

**Comment by xelan** [comment]
Great, thank you very much :+1:! I'm just thinking about the integration process for all tests. Maybe a good approach would be the creation of a feature-branch in the official repo where all test-related commits land first. Afterwards, this branch can be merged into master. What do you think?

Your PR looks good so far, just a few minor notes. And could you please squash your commits afterwards?

**Comment by iam-merlin** [comment]
done, I didn't include your php_cs... not sure if it's already used... lot of fix to do :cry: 

I can add a commented line in .travis.yml for php_cs if you want.

If it's ok, I will stash my commits.

File: tests/Helpers/UtilTest.php [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+/**
+ * This file is part of the php-resque package.
+ *
+ * (c) Michael Haynes <mike@mjphaynes.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Tests\Resque\Heplers\Utils;
```
Comment (xelan) [inline tests/Helpers/UtilTest.php]: I think the FQCN should be `Resque\Tests\Helpers\UtilTest` and it should be in the folder `tests/Helpers` to match the source folder structure. 

With this, autoload-rules can be set up to autoload the `Resque` namespace from `src` and autoload-dev the `Resque\Tests` namespace from `tests`. Using this approach, additional non-test classes (e.g. stub classes) can be autoloaded in development/testing, without the corresponding class loading overhead in production.

File: phpunit.xml.dist [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+        <env name="APP_ENV" value="test"/>
+        <env name="APP_DEBUG" value="1"/>
+    </php>
+
+    <testsuites>
+        <testsuite name="Resque Tests">
+            <directory>tests/</directory>
+        </testsuite>
+    </testsuites>
+</phpunit>
```
Comment (xelan) [inline phpunit.xml.dist]: Please add also a newline here, so files can be printed properly in the CLI.

File: tests/Helpers/UtilTest.php [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+/**
+ * Class UtilTest
+ *
+ * @author Romain DARY <romain.dary@eoko.fr>
+ */
+class UtilTest extends TestCase
+{
+    /**
+     * @param string $expected expected readable sizes
+     * @param integer $bytes size in bytes
```
Comment (xelan) [inline tests/Helpers/UtilTest.php]: Please use the shorthand type documentations (`int`/`bool`) to be in line with Symfony's code style and the proposed [PSR-5 standard](https://github.com/php-fig/fig-standards/blob/master/proposed/phpdoc.md).

**Comment by xelan** [comment]
Yeah, unfortunately some old code is not fully PSR-2 code-style compatible yet. The commented line in the Travis configuration sounds good for now.

**Comment by iam-merlin** [comment]
@xelan "your" php_cs is an unmaintened version and the code in php-resque is "not fully psr-2"  :/.

Can you add a more recent version of php-cs/php-cs-fixer and "auto fix" the project. The result is quite acceptable after testing.

The reason is to add php-cs to the CI.

**Comment by iam-merlin** [comment]
hi again,

I just add too php-cs-fix and I pushed to this branch all fixes :
- I rewrite your `.php_cs` to a v2 phpcs
- Add in composer.json some scripts (test, cs-ci, cs-fix)
- Fix all code (really need a review for that... I would prefer to let you lead this point)
- Update .travis
- Add blanck lines :P

**Comment by xelan** [comment]
Great, thank you @iam-merlin! I was on a business trip last week, sorry for not responding earlier.
My plan is now to review your changes this Friday. With all the CS improvements, the PR is quite large so it will take some time.

**Comment by iam-merlin** [comment]
no pb, CS if made through `composer cs-fix`, so it may be not perfect.

**Comment by xelan** [comment]
Thank you very much, @iam-merlin! I added some minor notes in particular concerning the namespace organization of the test suite.

Could you please have a look at that, rebase onto the current master, and squash your commits to one commit per part (test suite/code style improvements/toolchain upgrade)? This will make the further review process way easier.

Thank you very much, best regards!

**Comment by iam-merlin** [comment]
buh, I was so sure that I already fixed that xD. My mistake :/.

**Comment by iam-merlin** [comment]
voilà!

**Comment by iam-merlin** [comment]
last master change rebased + squash commits... should be ok

**Comment by xelan** [comment]
Thank you, @iam-merlin! Could you maybe just change/summarize the commit message as well? I think the fix typo etc are now not necessary anymore :smiley:
By the way, I think one comment (EOF line ending) is still unresolved.

I'll re-review tomorrow then. Thank you, best regards

File: phpunit.xml.dist (hunk 5 above)
Comment (iam-merlin) [reply phpunit.xml.dist]: done finally, my IDE make me crazy by showing and removing newline only on xml files...

**Comment by iam-merlin** [comment]
you can plug-in travis now :D 💃

**Comment by xelan** [comment]
Perfect 👍 I'll try that tomorrow.

File: composer.json [hunk 7]
```diff
@@ -25,15 +26,25 @@
         "ext-phpiredis": "Native PHP extension for Redis connectivity. Predis will automatically utilize when available."
     },
     "require-dev": {
-        "sami/sami": "~3.3|~4.0",
-        "phpunit/phpunit": "~4.8"
+        "phpunit/phpunit": "~4.8",
+        "friendsofphp/php-cs-fixer": "~2.2.0"
```
Comment (xelan) [inline composer.json]: I'm not sure whether the installation of php-cs-fixer in project scope is a good idea, as it adds a lot of dependencies and therefore possible side-effects. Is a global installation with composer possbile in Travis?

File: src/Resque/Helpers/CatchOutput.php [hunk 8]
```diff
@@ -27,8 +29,8 @@ class CatchOutput extends \Symfony\Component\Console\Output\Output
     /**
      * Constructor.
      *
-     * @param integer                       $verbosity The verbosity level (one of the VERBOSITY constants in OutputInterface)
-     * @param Boolean                       $decorated Whether to decorate messages
+     * @param int                       $verbosity The verbosity level (one of the VERBOSITY constants in OutputInterface)
```
Comment (xelan) [inline src/Resque/Helpers/CatchOutput.php]: Could you please change the alignment of the variables to start with one space after the longest type definition? Thank you 😄

File: src/Resque/Helpers/CatchOutput.php [hunk 9]
```diff
@@ -57,7 +59,7 @@ public function write($messages, $newline = false, $type = self::OUTPUT_RAW)
      * Stores message in a local string
      *
      * @param string  $message A message to write to the output
-     * @param Boolean $newline Whether to add a newline or not
+     * @param bool $newline Whether to add a newline or not
```
Comment (xelan) [inline src/Resque/Helpers/CatchOutput.php]: see above

File: src/Resque/Job.php [hunk 10]
```diff
@@ -330,7 +332,7 @@ public function perform()
 
             $this->complete();
 
-        // setUp said don't perform this job
+            // setUp said don't perform this job
```
Comment (xelan) [inline src/Resque/Job.php]: The comment should probably be aligned with the catch block, although php-cs-fixer complains... Maybe move it down?

File: src/Resque/Socket/Server.php [hunk 11]
```diff
@@ -187,12 +203,7 @@ public function run()
         }
 
         if (function_exists('pcntl_signal')) {
-            // PHP 7.1 allows async signals
-            if (function_exists('pcntl_async_signals')) {
-                pcntl_async_signals(true);
-            } else {
-                declare(ticks = 1);
-            }
+            declare(ticks = 1);
```
Comment (xelan) [inline src/Resque/Socket/Server.php]: Here the async feature (merged in #4ca7d32b) got lost. Could you please rebase on the current master to resolve this issue?

File: src/Resque/Worker.php [hunk 12]
```diff
@@ -306,7 +307,7 @@ public function work()
 
                 $this->shutdown();
 
-            // In parent if $pid > 0 since pcntl_fork returns process id of child
+                // In parent if $pid > 0 since pcntl_fork returns process id of child
```
Comment (xelan) [inline src/Resque/Worker.php]: see above

File: src/Resque/Worker.php [hunk 13]
```diff
@@ -505,12 +506,7 @@ public function register()
         if (function_exists('pcntl_signal')) {
             $this->log('Registering sig handlers for worker '.$this, Logger::DEBUG);
 
-            // PHP 7.1 allows async signals
-            if (function_exists('pcntl_async_signals')) {
-                pcntl_async_signals(true);
-            } else {
-                declare(ticks = 1);
-            }
+            declare(ticks = 1);
```
Comment (xelan) [inline src/Resque/Worker.php]: see above

**Comment by xelan** [comment]
Hi @iam-merlin!

A few minor things (well, one is a bit more, as it basically reverts the latest commit in master 😀) came up.

1. The async signal feature from 4ca7d32b is overwritten due to your changes, please rebase on the current master
2. Some comments can now be a bit misleading with changed indentation
3. After docblock changes, it would be nice to align the parameter variable names again - just that we're on it
4. Installation of php-cs-fixer in project scope is probably not the best idea, as it has a lot of dependencies. On [their website](https://github.com/FriendsOfPHP/PHP-CS-Fixer#installation), a global installation via phar or  `composer global require` is recommended.

 Thanks again and merry christmas

**Comment by iam-merlin** [comment]
:). Agreed for 3 first points, I will have a look.

For the last one, I'm really not a big fan of having external dev dependencies even if they like to show how to expose globally deps.

It's just dev deps and with that configuration, we don't have to read a guide to know of how to run test, etc. Just type `composer install && composer test` and you are ready to ship a perfect PR :P.

Eventually, good news, it's not necessary to save Octets anymore, the cost of the MegaOctet is really cheap nowadays :P

**Comment by xelan** [comment]
We can leave the `php-cs-fixer` in require-dev for now, and if there are any issues we need to think about it anyway.

**Comment by xelan** [comment]
Right, as there are many code style adaptations in this PR, merging other PRs based on an older master branch would be quite messy afterwards.

ping @iam-merlin - What do you think about the timeline for the completion of this PR? The Re-review will probably be quite quick afterwards.

Thank you again, happy new year :smile:

File: composer.json (hunk 7 above)
Comment (iam-merlin) [reply composer.json]: see comment above :D

File: src/Resque/Helpers/CatchOutput.php (hunk 8 above)
Comment (iam-merlin) [reply src/Resque/Helpers/CatchOutput.php]: done

File: src/Resque/Helpers/CatchOutput.php (hunk 9 above)
Comment (iam-merlin) [reply src/Resque/Helpers/CatchOutput.php]: done

File: src/Resque/Worker.php (hunk 12 above)
Comment (iam-merlin) [reply src/Resque/Worker.php]: done

File: src/Resque/Worker.php (hunk 13 above)
Comment (iam-merlin) [reply src/Resque/Worker.php]: done

File: src/Resque/Socket/Server.php (hunk 11 above)
Comment (iam-merlin) [reply src/Resque/Socket/Server.php]: done

**Comment by iam-merlin** [comment]
@xelan done, php-cs-fix is kept (one of your last comments).

**Comment by iam-merlin** [comment]
@scones ETA : 4 Jan 2018, 21:30 CET

**Comment by xelan** [comment]
Great, thanks :smile:
I'm doing a quick re-review of the PR now...

File: composer.json [hunk 14]
```diff
@@ -11,6 +11,7 @@
             "email": "mike@mjphaynes.com"
         }
     ],
+    "minimum-stability": "beta",
```
Comment (xelan) [inline composer.json]: Why do we need a minimum stability setting here? I could not find any untagged/unstable releases used.

File: src/Resque/Helpers/Util.php [hunk 15]
```diff
@@ -106,7 +108,10 @@ public static function human_time_diff($from, $to = null)
      * @param   mixed   key path string (delimiter separated) or array of keys
      * @param   mixed   value that was found
      * @param   string  key path delimiter
-     * @return  bool
+     * @param  mixed $array
```
Comment (xelan) [inline src/Resque/Helpers/Util.php]: Here the parameter documentation doesn't match anymore with the function signature - maybe a merge artifact?

File: src/Resque/Worker.php [hunk 16]
```diff
@@ -1249,6 +1249,7 @@ protected function updateProcLine($status)
 
     /**
      * Creates process title string from current version and status of worker
+     * @param mixed $status
```
Comment (xelan) [inline src/Resque/Worker.php]: This should probably be
```
@param string $status
@return string
```

**Comment by xelan** [comment]
Just a few minor changes, then we can merge it I'd say. Thank you very much again!

File: src/Resque/Helpers/Util.php (hunk 15 above)
Comment (scones) [reply src/Resque/Helpers/Util.php]: shouldn't the `mixed $array` be indented one more space to the right after the `@param`?

Comment (xelan) [reply src/Resque/Helpers/Util.php]: Oh, right the docblocks should be aligned as well after they are fixed here. Thanks :)

Comment (iam-merlin) [reply src/Resque/Helpers/Util.php]: autofix artifact, done

File: src/Resque/Worker.php (hunk 16 above)
Comment (iam-merlin) [reply src/Resque/Worker.php]: done

File: composer.json (hunk 14 above)
Comment (iam-merlin) [reply composer.json]: removed

**Comment by iam-merlin** [comment]
should be done

**Comment by iam-merlin** [comment]
here we are!

**Comment by xelan** [comment]
Wohooo, thank you very much Romain :smile:

**Comment by iam-merlin** [comment]
oh fuck, one last error... can you add in README.md travis badge (and ofc add travis integration) plus this fix :
```
diff --git i/src/Resque/Worker.php w/src/Resque/Worker.php
index 28cca97..19da728 100644
--- i/src/Resque/Worker.php
+++ w/src/Resque/Worker.php
@@ -1250,7 +1250,7 @@ class Worker
     /**
      * Creates process title string from current version and status of worker
      *
-     * @param string $status
+     * @param  string $status
      * @return string
      */
     protected function getProcessTitle($status)
```

**Comment by xelan** [comment]
No problem... I need to set up Travis anyway then I'll add the badge 😉

**Comment by scones** [comment]
as this pr is already merged into master, this should be a new issue / pr.

**Pull request merged**